### DISCUSSION TITLE
Using a mix of native and computer vision APIs

### DISCUSSION THREAD

**Discussion by 9nut** [AxisCommunications/discussions#158] [ACAP] [answered]
I would like to verify my understanding of the architecture when an application requires a mix of API's that need native and computer-vision runtimes. I would appreciate any feedback.

What I'm trying to do is roughly similar to what's outlined in `acap-computer-vision-sdk-examples/pose-estimator-with-flask` example. The app will use a specialized SSD to detect objects of interest and generate events for the RTSP stream which are then processed by a management and archiving system.  This would seem to require 4 containers:

1. Model container
2. Inference Server (tensorflow-serving gRPC)
3. App server container
4. Event handler container (something similar to `acap-native-sdk-examples/axevent/send_event`)

Is this correct?

If so, is there a protocol definition for Event API (to eliminate the 4th container)?

**Comment by 9nut** [comment]
If it matters, I'm signed up with Axis TIPP.

**Comment by Corallo** [comment, marked as answer]
Hello @9nut 

I can start making a bit of clarity, the example that you showed [pose-estimator-with-flask](acap-computer-vision-sdk-examples/pose-estimator-with-flask) is part of the computer vision SDK. While [send_event](acap-native-sdk-examples/axevent/send_event) of the native sdk.

The main difference, is that the first (CV sdk) runs containers in the camera, while the second, ACAP (.eap packaged applications).

I think it would be simpler to stick with one of the two, and you can do what you want to achieve with both.

If you for instance would like to use the CV SDK, you could just try to install an extra python package to produce an RTSP stream. I am not an expert on that, but on a quick search I saw that there is one called `imagezmq` that might do what you want.

**Reply by 9nut** [reply]
Thanks @Corallo 


To simplify my question: *Is there a way to use the Event API in an app performing inference in a CV container*?

I think the answer is no, because Event API and CV (tensorflow_serving) APIs have different runtime dependencies.

If that's not possible, then I would like to find an example that combines [object-detection](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/object-detection/app) with [send_event](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/axevent/send_event). Basically an example combining a glib loop to use the Event API to generate events, with a loop for using larod APIs to run inference on each frame).

**Reply by Corallo** [reply]
It is possible to have it in a container using the CV, but it is hard.
To use native APIs (like the event handler) in the CV examples, we rely on the [acap-runtime](https://github.com/AxisCommunications/acap-runtime), but the event_api is not currently included among the [available APIs](https://github.com/AxisCommunications/acap-runtime/tree/main/src). Since it is open source, you could try to include it by yourself, but might be hard.

If you need to use the event API it is probably simpler to use ACAP native examples as you said. We don't have an example that combine object-detection with send_event, but should be not too much work for you to add the libs in the send event example to the object detection. 

It should be enough, to got in the Makefile of the [object-detection example](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/object-detection/app/Makefile#L5) and replace the line
`PKGS = gio-2.0 vdostream gio-unix-2.0 liblarod` with
`PKGS = gio-2.0 vdostream gio-unix-2.0 liblarod glib-2.0 axevent`

Then you will be able to integrate one example in the other.

**Reply by 9nut** [reply]
Thank you @Corallo This is very helpful.
